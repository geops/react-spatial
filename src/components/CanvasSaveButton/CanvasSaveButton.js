import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import OLMap from 'ol/Map';
import { TiImage } from 'react-icons/ti';
import Button from '../Button';

const propTypes = {
  /**
   * Title of the button.
   */
  title: PropTypes.string.isRequired,

  /**
   *  Children content of the button.
   */
  children: PropTypes.element,

  /**
   * CSS class of the button.
   */
  className: PropTypes.string,

  /**
   * HTML tabIndex attribute
   */
  tabIndex: PropTypes.number,

  /**
   * Format to save the image.
   */
  saveFormat: PropTypes.oneOf(['image/jpeg', 'image/png']),

  /** An existing [ol/Map](https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html). */
  map: PropTypes.instanceOf(OLMap),
};

const defaultProps = {
  map: null,
  tabIndex: 0,
  children: <TiImage focusable={false} />,
  className: 'tm-canvas-save-button',
  saveFormat: 'image/png',
};

/**
 * This component displays a button to save canvas as an image.
 * It is a sub-component of the ShareMenu.
 */
class CanvasSaveButton extends PureComponent {
  constructor(props) {
    super(props);
    const { saveFormat } = this.props;
    this.options = { format: saveFormat };
    this.fileExt = this.options.format === 'image/jpeg' ? 'jpg' : 'png';
  }

  getDownloadImageName() {
    return (
      `${window.document.title.replace(/ /g, '_').toLowerCase()}` +
      `.${this.fileExt}`
    );
  }

  getCanvasImage(opts, asMSBlob) {
    const { map } = this.props;
    let image;

    map.once('postcompose', evt => {
      const { canvas } = evt.context;
      const { height, width } = canvas;

      const destinationCanvas = document.createElement('canvas');
      destinationCanvas.width = width;
      destinationCanvas.height = height;

      const destContext = destinationCanvas.getContext('2d');
      destContext.fillStyle = 'white';
      destContext.fillRect(0, 0, width, height);
      destContext.drawImage(canvas, 0, 0, width, height, 0, 0, width, height);

      if (asMSBlob) {
        image = destinationCanvas.msToBlob();
      } else {
        image = destinationCanvas.toDataURL(opts.format);
      }
    });
    map.renderSync();
    return image;
  }

  downloadCanvasImage(e) {
    if (/msie (9|10)/gi.test(window.navigator.userAgent.toLowerCase())) {
      // ie 9 and 10
      const w = window.open('about:blank', '');
      w.document.write(
        `<img src="${this.getCanvasImage(this.options)}" alt="from canvas"/>`,
      );
    } else if (window.navigator.msSaveBlob) {
      // ie 11 and higher
      window.navigator.msSaveBlob(
        new Blob([this.getCanvasImage(this.options, true)], {
          type: this.options.format,
        }),
        this.getDownloadImageName(),
      );
    } else {
      const link = document.createElement('a');
      link.download = this.getDownloadImageName();
      link.href = this.getCanvasImage(this.options);
      link.click();
    }

    if (window.navigator.msSaveBlob) {
      // ie only
      e.preventDefault();
      e.stopPropagation();
    }
  }

  render() {
    const { title, children, tabIndex, className } = this.props;

    return (
      <Button
        className={className}
        title={title}
        tabIndex={tabIndex}
        onClick={e => this.downloadCanvasImage(e)}
      >
        {children}
      </Button>
    );
  }
}

CanvasSaveButton.propTypes = propTypes;
CanvasSaveButton.defaultProps = defaultProps;

export default CanvasSaveButton;
