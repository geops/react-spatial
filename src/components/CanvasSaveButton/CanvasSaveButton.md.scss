.tm-canvas-save-button,
.tm-canvas-save-button:hover {
  color: white;

  svg {
    width: 75%;
  }
}
