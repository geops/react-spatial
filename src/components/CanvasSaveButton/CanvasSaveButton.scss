.tm-share-button,
.tm-share-button:hover {
  color: white;
}
