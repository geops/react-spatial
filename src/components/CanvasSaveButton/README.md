#

This demonstrates the use of CanvasSaveButton.

```jsx
import React from 'react';
import { TiImage } from 'react-icons/ti';
import Map from 'ol/Map';
import CanvasSaveButton from 'react-spatial/components/CanvasSaveButton';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import LayerService from 'react-spatial/LayerService';
import NorthArrow from '../NorthArrow/NorthArrow';
import Feature from 'ol/Feature';

const map = new Map();
const layers = ConfigReader.readConfig(treeData);
const layerService = new LayerService([...layers]);

function CanvasSaveButtonExample() {
  return (
    <div className="tm-canvas-save-button-example">
      <BasicMap
        map={map}
        center={[874105.13, 6106172.77]}
        zoom={10}
        layers={layers}
      />
      <CanvasSaveButton
        title="Save the map as PNG"
        className="tm-round-grey-hover-primary tm-button"
        map={map}
        extraData={{
          copyright: {
            text: () => {
              return layers[0].copyright;
            },
          },
          northArrow: {
            rotation: () => {
              return NorthArrow.radToDeg(map.getView().getRotation());
            },
            circled: true,
          },
        }}
      >
        <TiImage focusable={false} />
      </CanvasSaveButton>
    </div>
  );
}

<CanvasSaveButtonExample />;
```
