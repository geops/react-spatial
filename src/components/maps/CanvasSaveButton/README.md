#

This demonstrates the use of CanvasSaveButton.

```jsx
import React from 'react';
import { TiImage } from 'react-icons/ti';
import Map from 'ol/Map';
import degrees from 'radians-degrees';
import CanvasSaveButton from 'react-spatial/components/maps/CanvasSaveButton';
import BasicMap from 'react-spatial/components/maps/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import LayerService from 'react-spatial/LayerService';
import NorthArrow from '../NorthArrow/NorthArrow';
import Feature from 'ol/Feature';

const map = new Map();
const layers = ConfigReader.readConfig(treeData);
const layerService = new LayerService([...layers]);

function CanvasSaveButtonExample() {
  return (
    <div className="rs-canvas-save-button-example">
      <BasicMap
        map={map}
        center={[874105.13, 6106172.77]}
        zoom={10}
        layers={layers}
        tabIndex={0}
      />
      <CanvasSaveButton
        map={map}
        extraData={{
          copyright: {
            text: () => {
              return layers[0].copyright;
            },
          },
          northArrow: {
            rotation: () => {
              return degrees(map.getView().getRotation());
            },
            circled: true,
          },
        }}
      >
        <TiImage focusable={false} />
      </CanvasSaveButton>
    </div>
  );
}

<CanvasSaveButtonExample />;
```
