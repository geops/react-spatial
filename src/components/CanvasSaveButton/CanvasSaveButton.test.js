import 'jest-canvas-mock';
import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import OLMap from 'ol/Map';
import OLView from 'ol/View';
import { TiImage } from 'react-icons/ti';
import CanvasSaveButton from './CanvasSaveButton';

configure({ adapter: new Adapter() });

describe('CanvasSaveButton', () => {
  const conf = {
    title: 'Karte als Bild speichern.',
    icon: <TiImage focusable={false} />,
    className: 'ta-example',
    saveFormat: 'image/jpeg',
  };
  const olView = new OLView();
  const olMap = new OLMap({ view: olView });

  test('should match snapshot.', () => {
    const component = renderer.create(
      <CanvasSaveButton
        title={conf.title}
        saveFormat={conf.saveFormat}
        map={olMap}
      >
        {conf.icon}
      </CanvasSaveButton>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should be trigger click function.', () => {
    const wrapper = shallow(
      <CanvasSaveButton
        className="ta-example"
        title={conf.title}
        saveFormat={conf.saveFormat}
        map={olMap}
      >
        {conf.icon}
      </CanvasSaveButton>,
    );
    const spy = jest.spyOn(CanvasSaveButton.prototype, 'downloadCanvasImage');

    wrapper.find('.ta-example').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
