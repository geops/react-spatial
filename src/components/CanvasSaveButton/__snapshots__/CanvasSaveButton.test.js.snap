// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasSaveButton should match snapshot. 1`] = `
<div
  aria-label="Karte als Bild speichern."
  className="tm-canvas-save-button"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="Karte als Bild speichern."
>
  <svg
    baseProfile="tiny"
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    version="1.2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="8.5"
      cy="8.501"
      r="2.5"
    />
    <path
      d="M16 10c-2 0-3 3-4.5 3s-1.499-1-3.5-1c-2 0-3.001 4-3.001 4h14.001s-1-6-3-6zM20 3h-16c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h16c1.103 0 2-.897 2-2v-12c0-1.103-.897-2-2-2zm0 14h-16v-12h16v12z"
    />
  </svg>
</div>
`;
